docs/design/extending-api.md

```diff
@@ -0,0 +1,191 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/design/extending-api.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Extending the Kubernetes API
+
+This document describes the design for implementing the storage of custom API types in the Kubernetes API Server.
+
+
+## Object Model
+
+### The ThirdPartyObject
+
+The `ThirdPartyObject` resource describes the multiple versions of a custom object that the user wants to add
+to the Kubernetes API.
+
+Each `ThirdPartyObject` resource has the following:
+   * Standard Kubernetes object metadata.
+   * Description - A free text description of the object.
+   * APIGroup - An API group that this object should be placed into.
+   * Versions - One or more `Version` objects.
+
+### The `Version` Object
+
+The `Version` object describes a single concrete version of a custom object.  The `Version` object currently
+only specifies a single `Name` of the version.  It is kept as a separate object so that it can
+be additively expanded in the future to enable validation and other extensions.
+
+## Expectations about third party objects
+
+Every object that is added to a third-party Kubernetes object store is expected to contain Kubernetes
+compatible [object metadata](../devel/api-conventions.md#metadata).  This requirement enables the
+Kubernetes API server to provide the following features:
+   * Filtering lists of objects via LabelQueries
+   * `resourceVersion`-based optimistic concurrency via compare-and-swap
+   * Versioned storage
+   * Event recording
+   * Integration with basic `kubectl` command line tooling.
+
+## Usage
+
+When a user creates a new `ThirdPartyObject`, the Kubernetes API Server reacts by creating a new, namespaced
```

Comment:
Yes, namespaced only for now, I don't see many usecases for non-namespaced custom objects.  Added a note to this effect.